openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PatientResourceImpl.java

```diff
@@ -83,6 +87,27 @@ public String getMotechPatientIdentifierUuid() throws HttpException {
     }
 
     @Override
+    public String getPatientIdentifierName(String uuid) throws HttpException {
+        String identifierName = identifierUuidByName.get(uuid);
+
+        if (identifierName == null) {
+            PatientIdentifierListResult result = getAllPatientIdentifierTypes();
+            for (IdentifierType type : result.getResults()) {
+                if (StringUtils.equals(uuid, type.getUuid())) {
+                    if (openmrsInstance.getPatientIdentifierTypesNames().contains(type.getName())) {
+                        identifierName = type.getName();
```

Comment:
I know that it is complicated and it's not easy to understand what happens in this module. I will try to explain you how it works and maybe you can suggest how to improve understandability of this code. Personally, I assume that we should write a documentation for OpenMRS.

1. In OpenMRS we can define Patient Identifier Types, for example : MOTECH ID, CommCare CaseID, OpenMRS ID. 
2. Every Identifier Type has unique name and uuid, which is generated by OpenMRS.
3. When we create Patient, we can add Identifiers giving name of the Identifier Type and identifier number.

In this method we retrieve the Patient Identifier Type name (OpenMRS ID, CommCare CaseID etc.) giving Uuid.

After we have the name of Identifier Type, we can add an Identifier to Patient.  Identifier object looks like the following :
- String identifier (value, number for Patient identifier, for example 100251)
- IdentifierType identifierType (object with identifier name, for example : MOTECH ID)

So answering to your question, we don't retrieve identifier type name by identifier value, but by their uuid. Here there is no connection with concrete Patient object.

I hope, that it is clear to you. Maybe I can write this method better or extend javadoc. Let me know what do you think about it.